vaadin-spring/src/main/java/com/vaadin/flow/spring/DevModeBrowserLauncher.java

```diff
@@ -0,0 +1,130 @@
+
+import com.vaadin.base.devserver.util.BrowserLauncher;
+import com.vaadin.flow.server.VaadinContext;
+import com.vaadin.flow.server.VaadinServletContext;
+import com.vaadin.flow.server.startup.ApplicationConfiguration;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.SpringApplicationRunListener;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.web.context.support.GenericWebApplicationContext;
+
+/**
+ * Utilities for launching a browser when running in development mode.
+ */
+public class DevModeBrowserLauncher
+        implements SpringApplicationRunListener, Serializable {
+
+    private static final String LAUNCH_TRACKER = "LaunchUtil.hasLaunched";
+    private static final String LAUNCHED_VALUE = "yes";
+
+    public DevModeBrowserLauncher(SpringApplication application,
+            String[] arguments) {
+    }
+
+    @Override
+    public void running(ConfigurableApplicationContext context) {
+        try {
+            VaadinConfigurationProperties properties = context
+                    .getBean(VaadinConfigurationProperties.class);
+
+            if (properties.isLaunchBrowser()) {
+                launchBrowserInDevelopmentMode(context);
+            }
+        } catch (Exception e) {
+            getLogger().debug("Failed to launch browser", e);
+        }
+    }
+
+    /**
+     * Launch the default browser and open the given application base URL if
+     * running in development mode.
+     *
+     * Does nothing if the application is running in production mode.
+     * 
+     * @param applicationContext
+     *            the application context
+     */
+    private void launchBrowserInDevelopmentMode(ApplicationContext appContext) {
+        if (isLaunched()) {
+            // Only launch browser on startup, not on reload
+            return;
+        }
+        if (!(appContext instanceof GenericWebApplicationContext)) {
+            getLogger().warn(
+                    "Unable to determine production mode for an Spring Boot application context of type "
+                            + appContext.getClass().getName());
+            return;
+        }
+        GenericWebApplicationContext webAppContext = (GenericWebApplicationContext) appContext;
+        if (!DevModeBrowserLauncher.isProductionMode(webAppContext)) {
+            String location = getUrl(webAppContext);
+            String outputOnFailure = "Application started at " + location;
+            try {
+                BrowserLauncher.launch(location, outputOnFailure);
+                setLaunched();
+            } catch (Exception | NoClassDefFoundError e) { // NOSONAR
+                // NoClassDefFoundError happens if vaadin-dev-server is not on
+                // the classpath
+                getLogger().info(outputOnFailure);
```

Comment:
1. If `vaadin-dev-server` is not on the classpath, then at least `NoClassDefFoundError` should be logged, or even better be thrown telling to the developer that it's a mistake and this dependency should be added.
2. IMO `Exception` doesn't need to be caught here, otherwise it wouldn't be clear what happened with the application, but instead only  `Application started at ...` would be shown, which gives no idea what was the original error and how to fix it.

Reply:
The idea with catching all exceptions is not to bother the user if a browser window couldn't be opened automatically. It is not really a problem with the project and if it doesn't work, you do not want a long exception logged every time you start the server.

Reply:
I also have a feeling that somebody has decided that you should be able to run in development mode without vaadin-dev-server but I do not know why. Am I just wrong @pleku and we should log `NoClassDefFoundError`? (I hope I am)

Reply:
> The idea with catching all exceptions is not to bother the user if a browser window couldn't be opened automatically. It is not really a problem with the project and if it doesn't work, you do not want a long exception logged every time you start the server.

Okay, I see. So if some exception is thrown and launching the browser fails, then this message is logged giving the URL to developer to click on to open the browser manually. Sounds reasonable.

Reply:
It has been possible since v10 to run in not-production-mode but use the production built bundle. It was even requested at some point. Probably not used 99 % of the time but I don't know why we would prevent it now if it is not a significant effort to keep it possible.